torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -324,9 +325,39 @@ void ConvertGraphToONNXProto(
   }
 }
 
+// this function checks wheather the blocks of If node have the same return
+// type.
+bool IsBlockReturnTypeSame(Node* n) {
+  TORCH_INTERNAL_ASSERT(n->kind() == ::c10::onnx::If);
+  auto then_block = n->blocks()[0];
+  auto else_block = n->blocks()[1];
+  for (size_t i = 0; i < n->outputs().size(); i++) {
+    // check the type
+    auto then_block_type = then_block->outputs()[i]->type();
+    auto else_block_type = else_block->outputs()[i]->type();
+    if (then_block_type->cast<TensorType>() &&
+        else_block_type->cast<TensorType>()) {
+      if (then_block_type->cast<TensorType>()->scalarType() !=
+          else_block_type->cast<TensorType>()->scalarType()) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
 // Any additional post process that are specific to individual node kind.
 void SpecialPostProcess(Node* n) {
   switch (n->kind()) {
+    case ::c10::onnx::If: {
+      if (!IsBlockReturnTypeSame(n) && IsStaticConditionONNX(n)) {
+        auto cond = ConditionValueONNX(n) ? 0 : 1;
```

Comment:
nit: variable name cond is misleading (value is opposite), consider David's suggestion